### ISSUE TITLE
TypeError: Cannot read property 'req' of undefined

### ISSUE THREAD

**Issue by andreynazarov3** [nuxt-community/nuxt-i18n#711]
### Version
nuxt-i18n:  6.11.1
nuxt: 2.11.0

#### [mode]
 - [ x ] universal

### Nuxt-i18n configuration
```
i18n: {
    strategy: 'prefix_except_default',
    locales: [
      { code: 'en', file: 'en.json' },
      { code: 'de', file: 'de.json' }
    ],
    defaultLocale: 'en',
    vueI18n: {
      fallbackLocale: 'en'
    },
    lazy: true,
    langDir: 'i18n/',
    detectBrowserLanguage: {
      useCookie: true,
      cookieKey: 'i18n_redirected',
      fallbackLocale: 'en'
    }
  }
```

### Steps to reproduce

Try to invoke `localeLink` function on `$root` on SSR. Like
```
const { $root } = this;
$root.localePath(...)
```

### What is Expected?

Path is generated properly.

### What is actually happening?

Getting TypeError.

It is known issue been discussed in ticket https://github.com/nuxt-community/nuxt-i18n/pull/596
And the PR was merged, but now looks like it has been reverted...
In a nutshell, on SSR `$root` instance `$ssrContext` should be replaced by `ssrContext`.

**Comment by rchl** [comment]
#596 PR was never merged as there was no reproducible testcase.
Why do you need to get `$root` instead of using `this.localePath` directly?

**Comment by andreynazarov3** [comment]
> #596 PR was never merged as there was no reproducible testcase.
> Why do you need to get `$root` instead of using `this.localePath` directly?

In my case I was trying to use `$root` because for some reason `this` was undefined in some cases. I am guessing that sometimes component get destroyed before `localePath` executes. So this is making safer use of `localePath`.

**Comment by rchl** [comment]
I'd like to see a specific code that triggers it.

If you run an async code from `created` hook, for example, it might happen that instance will be destroyed while your code is running. In that case `this` won't be null but it will point to a destroyed instance and likely crash when trying to use some API from it.

If you incorrectly bind your method handler (for example by using arrow functions where you shouldn't) then `this` might be wrong or undefined.

But in both cases, you should refactor the code to account for this rather than coming up with a workarounds that can cause issues later.

So unless I see a specific code that makes sense and triggers the issue, I'm not just gonna blindly replace `$ssrContext` with `ssrContext`.

**Comment by rchl** [comment]
And especially in SSR that would indicate a bad code as there are not that many places where you can run your async code and can't await the result.

**Comment by andreynazarov3** [comment]
well, the code I have is simple, i have a mixin with a methods, and one of them is this
```
methods: {
    getLink(slug) {
      return this.localePath({
        name: 'product-slug',
        params: { slug }
      });
    }
}
```
It's usually working, but some users experience `this is undefined` here.
To be more exact, error happens inside localePath function
```
null is not an object (evaluating 'n.locale')
```

**Comment by rchl** [comment]
Well, that doesn't look like something that would be triggered in SSR though.

And if it would, I don't see how the component would get destroyed before calling it. In SSR components don't get destroyed normally, only created.

But there is nothing wrong with this code itself either way, the issue would be with the code calling it.

I can see how this could be a potential issue on the client-side where it's harder to ensure that component exists when you are calling this method. But if you are using @click handler, it should work reliably. If you are setting DOM listeners yourself or calling this from async functions then it might be problematic.

But in any case, the [latest version of vue-i18n](https://github.com/kazupon/vue-i18n/releases/tag/v8.17.6) is no longer destroying the `$i18n` instance from within `beforeDestroy` so it should fix your issue. If you uninstall `nuxt-i18n` and install it again you should get it updated to latest version.

**Comment by andreynazarov3** [comment]
Okay, thanks, I will try to update vue-i18n and see what happens. 

Error that I described is happening only client-side, so I am guessing that it caused by destroyed component.

**Comment by rchl** [comment]
Resolving as:
 - it's not likely a problem in nuxt-i18n
 - shouldn't happen with new vue-i18n

**Comment by lukeromanowicz** [comment]
If you are using vue-composition-api plugin it's because root's $ssrContext is undefined. It's moved to the context.ssrContext. I am not sure if the same case happens to non CAPI scenarios

**Comment by dschreij** [comment]
@lukeromanowicz thanks for that. I am using the composition api plugin (from @nuxtjs/composition-api to be specific). Do you have any workaround for this issue? Does using the 'next' version of this module solve this problem?

**Comment by rchl** [comment]
I have not investigated composition api yet and there is no `next` version.

**Comment by dschreij** [comment]
> and there is no next version.

Apologies, I think I accidentally confused this with vue-i18n itself, in which they have added support for capi if I'm not mistaken.
